File: Softeer/Lv3/효도 여행.py
import sys
from collections import defaultdict, deque

n, m = map(int, sys.stdin.readline().strip().split())
string = sys.stdin.readline().strip()
node_dict = defaultdict(set)
for _ in range(n - 1):
    command = sys.stdin.readline().strip().split()
    node_dict[int(command[0])].add((int(command[1]), command[2]))
    node_dict[int(command[1])].add((int(command[0]), command[2]))

def solution():
    answer = 0
    q = deque([(1, [0] * (m + 1), '', set([1]))])
    while q:
        node, dp, prev_str, visited = q.popleft()
        for n_node, char in node_dict[node]:
            if n_node in visited:
                continue
            visited.add(n_node)
            new_dp = [0] * (m + 1)
            for i in range(1, m + 1):
                if char == string[i - 1]:
                    new_dp[i] = dp[i - 1] + 1
                else:
                    new_dp[i] = max(dp[i], new_dp[i - 1])
                answer = max(answer, new_dp[i])
            q.append((n_node, new_dp, prev_str + char, visited))
    
    return answer

print(solution())

File: Softeer/Lv3/효도 여행.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const [n, m] = input[0].split(' ').map(item => Number(item));
const string = input[1];
const nodeDict = {};
for(let i = 0; i < n - 1; i++) {
    const arr = input[i + 2].split(' ');
    const a = Number(arr[0]);
    const b = Number(arr[1]);
    const char = arr[2];
    if(a in nodeDict) {
        nodeDict[a].push([b, char]);
    } else {
        nodeDict[a] = [[b, char]];
    }
    if(b in nodeDict) {
        nodeDict[b].push([a, char]);
    } else {
        nodeDict[b] = [[a, char]];
    }
}

function solution() {
    let answer = 0;

    let visited = new Set([1])
    let q = [[1, new Array(m + 1).fill(0)]];
    while(q.length > 0) {
        let [node, dp] = q.shift();
        answer = Math.max(answer, dp[dp.length - 1])
        for(let i = 0; i < nodeDict[node].length; i++) {
            const [nNode, nChar] = nodeDict[node][i];
            if(visited.has(nNode)) {
                continue;
            }
            visited.add(nNode);
            let newDp = new Array(m + 1).fill(0);
            for(let j = 0; j < m; j++) {
                if(string[j] === nChar) {
                    newDp[j + 1] = dp[j] + 1;
                }
                else {
                    newDp[j + 1] = Math.max(dp[j + 1], newDp[j]);
                }
            }
            q.push([nNode, newDp])
        }
    }
    
    return answer;
}

console.log(solution());
